File: 085_Binary Tree Zigzag Level Order Traversal.py
"""
Binary Tree Zigzag Level Order Traversal
Return zigzag level order (left-to-right then right-to-left alternation).
"""
from typing import Optional, List

class TreeNode:
    def __init__(self,val=0,left=None,right=None):
        self.val=val; self.left=left; self.right=right

def zigzag_level_order(root: Optional[TreeNode]) -> List[List[int]]:
    if not root: return []
    from collections import deque
    q = deque([root]); res=[]
    left_to_right = True
    while q:
        n = len(q); row = [0]*n
        for i in range(n):
            node = q.popleft()
            idx = i if left_to_right else (n-1-i)
            row[idx] = node.val
            if node.left: q.append(node.left)
            if node.right: q.append(node.right)
        res.append(row)
        left_to_right = not left_to_right
    return res

if __name__ == '__main__':
    root = TreeNode(3)
    root.left = TreeNode(9)
    root.right = TreeNode(20)
    root.right.left = TreeNode(15)
    root.right.right = TreeNode(7)
    for row in zigzag_level_order(root): print(row)

# Time Complexity: O(n)
# Space Complexity: O(n)

File: 085_Binary Tree Zigzag Level Order Traversal.cpp
// Binary Tree Zigzag Level Order Traversal
// Return zigzag level order traversal (alternating directions per level).
#include <bits/stdc++.h>
using namespace std;
struct TreeNode{int val; TreeNode* left; TreeNode* right; TreeNode(int x): val(x), left(nullptr), right(nullptr) {}};

vector<vector<int>> zigzagLevelOrder(TreeNode* root){
    vector<vector<int>> res;
    if(!root) return res;
    queue<TreeNode*> q; q.push(root);
    bool leftToRight = true;
    while(!q.empty()){
        int n=q.size(); vector<int> row(n);
        for(int i=0;i<n;++i){ TreeNode* t=q.front(); q.pop(); int idx = leftToRight? i : (n-1-i); row[idx]=t->val; if(t->left) q.push(t->left); if(t->right) q.push(t->right);} 
        res.push_back(row);
        leftToRight = !leftToRight;
    }
    return res;
}

int main(){
    TreeNode* root = new TreeNode(3);
    root->left = new TreeNode(9);
    root->right = new TreeNode(20);
    root->right->left = new TreeNode(15);
    root->right->right = new TreeNode(7);
    auto v = zigzagLevelOrder(root);
    cout<<"Zigzag:\n";
    for(auto &r:v){ for(int x:r) cout<<x<<" "; cout<<"\n"; }
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(n)
